### PR TITLE
[llvm-jitlink] Support plain AArch32 range extension stubs in jitlink-check's stub_addr() expressions

### PR DESCRIPTION
**Pull request by weliveindetail**
We want to use regular `stub_addr()` expressions in `jitlink-check` lines to test the generation of stubs in AArch32, but we don't want this to require a standardized GOT-based PLT implementation. In terms of performance and binary size it doesn't seem beneficial. And in terms of patching, we should be able to handle range-extension- and interworking-stubs without a lot of extra logic.

This patch adds a separate path for AArch32 stubs in `llvm-jitlink-elf` to allow it. The relocations in our stubs are not pointing to the GOT, but to the external symbol directly. Thus, we have to avoid access to the block of the edge target. Instead we only return the symbol name. This is enough to use `stub_addr()` expressions in tests.

In order to allow decoding of stub target addresses in the future, we mention the stub flavor in the section name.

### CODE REVIEW

**Review by lhames — APPROVED** [review]
LGTM. Thanks Stefan!

File: llvm/tools/llvm-jitlink/llvm-jitlink-elf.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
 #include "llvm-jitlink.h"
 
+#include "llvm/ExecutionEngine/Orc/Core.h"
```
Comment (eymay) [inline llvm/tools/llvm-jitlink/llvm-jitlink-elf.cpp:15]: I think this can be removed

File: llvm/test/ExecutionEngine/JITLink/AArch32/ELF_thumb_stubs.s
Comment (eymay) [inline llvm/test/ExecutionEngine/JITLink/AArch32/ELF_thumb_stubs.s:1]: We might play with section name here, in the future possibly.

**Review by eymay — APPROVED** [review]
My understanding is that now in JITLink AArch32 we can have PLT and interworking/range-extension stubs in different sections so they are kept orthogonal features. Now the checker is also acknowledging this separation.

File: llvm/test/ExecutionEngine/JITLink/AArch32/ELF_thumb_stubs.s
Comment (weliveindetail) [reply llvm/test/ExecutionEngine/JITLink/AArch32/ELF_thumb_stubs.s:1]: Well, ideally I'd like to have `jitlink-check` lines like this:
```
# jitlink-check: decode_operand(stub_addr(elf_stubs.o, fn) + 0, 1) = (fn & 0x0000ffff)
# jitlink-check: decode_operand(stub_addr(elf_stubs.o, fn) + 4, 2) = (fn & 0xffff0000) >> 16
```

We know it's a Thumv7 stub and we could check the `movw` + `movt` immediates directly, but `decode_operand()` doesn't support address expressions right now. We could add that, but it doesn't seem trivial.

One alternative might be to do it in the checker itself and then add something like:
```
# jitlink-check: decode_stub_target(elf_stubs.o, fn) = fn
```
In this case the section name would determine the stub type for the checker. Not urgent yet, but something to think about.

File: llvm/tools/llvm-jitlink/llvm-jitlink-elf.cpp (hunk 1 above)
Comment (weliveindetail) [reply llvm/tools/llvm-jitlink/llvm-jitlink-elf.cpp]: Thanks, yes it's a leftover. Together with the `orc::ExecutionSession &ES` parameter below...

**Review by weliveindetail — COMMENTED** [review]
> My understanding is that now in JITLink AArch32 we can have PLT and interworking/range-extension stubs in different sections so they are kept orthogonal features. Now the checker is also acknowledging this separation.

Yes, I should express that in the commit message. Thanks for your notes!

**Pull request merged**